### PR TITLE
[sdk-gen] Don't JSON-encode provider config

### PR DESCRIPTION
**Pull request by blampe**
# Description

Most of our SDKs (node, python, dotnet, java) send JSON-encoded provider configuration for legacy reasons, but YAML doesn't do this ([example](https://github.com/pulumi/pulumi-go-provider/pull/172#issuecomment-1874600691)).

This behavior originated a long time ago when our providers only supported [string properties](https://github.com/pulumi/pulumi-terraform-bridge/commit/b4446913ae60dcce95c1bdd0b0d85d119f2c91c8), but of course this is no longer the case. Fortunately providers today accept both JSON or strongly typed configuration -- they attempt vanilla `plugin.UnmarshalProperties` but will [fall back](https://github.com/pulumi/pulumi-terraform-bridge/blob/bda33f369ae3e8f69cae5d7286e6ff35e0967964/pkg/tfbridge/config_encoding.go#L139-L156) to JSON deserialization if that fails. This explains why YAML is able to send rich configuration without issue, and why it's safe to remove the legacy JSON serialization from other SDKs.

The only exception is the newer go-provider, which doesn't handle non-primitive config from our SDKs (only from YAML https://github.com/pulumi/pulumi-go-provider/issues/171) because it doesn't do any JSON unmarshalling. We could either:
1. add this legacy JSON unmarshalling to go-provider, and support it indefinitely in the same way tfbridge will need to support it forever (because JSON config will be persisted in state files), or
2. stop JSON encoding config from SDKs, so go-provider only needs to concern itself with rich configuration.

This PR opts for (2) and removes JSON serialization from provider configs. https://github.com/pulumi/pulumi-java/pull/1308 makes a similar change for java.

Fixes https://github.com/pulumi/pulumi/issues/12773
Fixes https://github.com/pulumi/pulumi-go-provider/issues/171
Supersedes https://github.com/pulumi/pulumi-go-provider/pull/172
Related https://github.com/pulumi/pulumi-java/pull/1308
Related https://github.com/pulumi/pulumi-terraform-bridge/pull/1610
Refs https://github.com/pulumi/pulumi/issues/13435
Refs https://github.com/pulumi/pulumi/pull/7058
Refs https://github.com/pulumi/pulumi-yaml/issues/455
Refs https://github.com/pulumi/pulumi-docker/issues/456

## Checklist

- [ ] I have run `make tidy` to update any new dependencies
- [ ] I have run `make lint` to verify my code passes the lint check
  - [ ] I have formatted my code using `gofumpt`

### CODE REVIEW

**Comment by Frassle** [comment]
Looks like this hasn't been needed since Pulumi 1.1, not sure why we didn't take it out before? Probably worth having an old timer like @justinvp or @pgavlin to sanity check this, there may be some subtle reason its been left in for so long.

**Comment by t0yv0** [comment]
This is very tempting. Obviously feel like we must depend on it somewhere but if we don't that's a great change. QQ, when passing "variables" in Configure, they're still JSON-encoded and this PR does not affect that right?  For bridged providers we've weaned off reading `variables` but I think some natives still do. Perhaps that's handled in the engine not here in codegen? 

```
message ConfigureRequest {
    map<string, string> variables = 1; // a map of configuration keys to values.
    google.protobuf.Struct args = 2;   // the input properties for the provider. Only filled in for newer providers.
    bool acceptSecrets  = 3;           // when true, operations should return secrets as strongly typed.
    bool acceptResources = 4;          // when true, operations should return resources as strongly typed values to the provider.
    bool sends_old_inputs = 5;         // when true, diff and update will be called with the old outputs and the old inputs.
    bool sends_old_inputs_to_delete = 6; // when true, delete will be called with the old outputs and the old inputs.
}

```

**Comment by Frassle** [comment]
> Perhaps that's handled in the engine not here in codegen?

It is handled in the engine, was changed all the way back in pulumi 1.1.

**Comment by t0yv0** [comment]
Anything I can help with to unblock AWS  tests?

**Comment by blampe** [comment]
> Anything I can help with to unblock AWS tests?

@t0yv0 no need, I expect things will be green after https://github.com/pulumi/pulumi-awsx/pull/1201 goes in. The only other issues were a small type error in the dotnet fix and https://github.com/pulumi/pulumi-awsx/pull/1198.

**Comment by t0yv0** [comment]
This looks fantastic. Can we loop this through downstream bridged provider tests just in case to double-check that the bridge logic holds muster to parse these simpler config forms?

**Comment by t0yv0** [comment]
@justinvp any chance this can get a bit of review love? Also, how can we run the downstream tests to check bridged provider builds against this change, that'd help confidence levels. Much appreciated!

**Comment by justinvp** [comment]
> Fortunately providers today accept both JSON or strongly typed configuration -- they attempt vanilla `plugin.UnmarshalProperties` but will [fall back](https://github.com/pulumi/pulumi-terraform-bridge/blob/bda33f369ae3e8f69cae5d7286e6ff35e0967964/pkg/tfbridge/config_encoding.go#L139-L156) to JSON deserialization if that fails.

The native providers do this as well?

**Comment by t0yv0** [comment]
That's an excellent question, that's also not covered by automatic downstream tests. I think last time I looked at azure-native it was reading from deprecated "variables" instead of "args" that is necessarily JSON-encoded. Presumably this should not be affected by this change but nice to confirm tomorrow.

File: pkg/codegen/nodejs/gen.go [hunk 1]
```diff
@@ -790,11 +790,10 @@ func (mod *modContext) genResource(w io.Writer, r *schema.Resource) (resourceFil
...
 				}
+			}
 
-				// provider properties must be marshaled as JSON strings.
-				if r.IsProvider && !isStringType(prop.Type) {
-					arg = fmt.Sprintf("pulumi.output(%s).apply(JSON.stringify)", arg)
-				}
+			if r.IsProvider && !isStringType(prop.Type) {
+				arg = fmt.Sprintf("pulumi.output(%s)", arg)
 			}
```
Comment (justinvp) [inline pkg/codegen/nodejs/gen.go:797]: I don't think we need to keep this. We were only wrapping in `pulumi.output` so we could call `apply`.

File: pkg/codegen/python/gen.go [hunk 2]
```diff
@@ -1374,16 +1374,13 @@ func (mod *modContext) genResource(res *schema.Resource) (string, error) {
...
 		if res.IsProvider && !isStringType(prop.Type) {
 			if prop.Secret {
-				arg = fmt.Sprintf("pulumi.Output.secret(%s).apply(pulumi.runtime.to_json) if %s is not None else None", arg, arg)
+				arg = fmt.Sprintf("pulumi.Output.secret(%s) if %s is not None else None", arg, arg)
 				handledSecret = true
 			} else {
-				arg = fmt.Sprintf("pulumi.Output.from_input(%s).apply(pulumi.runtime.to_json) if %s is not None else None", arg, arg)
+				arg = fmt.Sprintf("pulumi.Output.from_input(%s) if %s is not None else None", arg, arg)
 			}
 		}
```
Comment (justinvp) [inline pkg/codegen/python/gen.go:1385]: I think all of this can go for the same reason.

**Review by justinvp — COMMENTED** [review]
Added some initial comments. Still want to dig into this a bit more.

Do we know what happens with existing state that's json encoded? I assume reading that state will continue to come back json encoded? Which may be problematic when enabling https://github.com/pulumi/pulumi/issues/13435.

File: pkg/codegen/nodejs/gen.go (hunk 1 above)
Comment (blampe) [reply pkg/codegen/nodejs/gen.go:797]: Yeah, I was intentionally trying to scope this to the absolute bare minimum changes but I will try pulling this out.

**Review by blampe — COMMENTED** [review]
@justinvp

> > The native providers do this as well?

> That's an excellent question, that's also not covered by automatic downstream tests. I think last time I looked at azure-native it was reading from deprecated "variables" instead of "args" that is necessarily JSON-encoded. Presumably this should not be affected by this change but nice to confirm tomorrow.

As far as I can tell, _all_ of our native providers are still using deprecated `GetVariables`, presumably because they don't have access to the bridge's transcoding logic.

* AWS: https://github.com/pulumi/pulumi-aws-native/blob/c3f606904a17fc1b484fb720ed0e306f6d0d3703/provider/pkg/provider/provider.go#L287
* Kubernetes: https://github.com/pulumi/pulumi-kubernetes/blob/3c6513ebe85930940b4b20e27f1b2cdcbd411f6e/provider/pkg/provider/provider.go#L455
* Azure: https://github.com/pulumi/pulumi-azure-native/blob/1bc180b9fc21de370637ee0c48eeea11fe62d18b/provider/pkg/provider/provider.go#L177
* GCP: https://github.com/pulumi/pulumi-google-native/blob/91279b497d3de878773059b7a5ed0063353146ba/provider/pkg/provider/provider.go#L114
* Pulumi Cloud: https://github.com/pulumi/pulumi-pulumiservice/blob/8083e3e00e02932a0b54d849f46a21eecf924f13/provider/pkg/provider/provider.go#L106 

This is both good and bad. Bad because it's a lot of cruft to clean up, but _mostly_ good because it means we didn't allow JSON encoding to proliferate beyond the bridge. That reduces the impact of this PR (since we won't affect native providers) and clears the way for modernizing native config.

> Do we know what happens with existing state that's json encoded? I assume reading that state will continue to come back json encoded? Which may be problematic when enabling https://github.com/pulumi/pulumi/issues/13435.

Existing state is OK regardless of whether it's JSON or not -- the bridge will continue to transcode JSON, and that logic will need to remain forever for compatibility.

The question of what happens to state (JSON or not) after we read it depends on https://github.com/pulumi/pulumi-terraform-bridge/pull/1610. Without that change we'll continue persisting JSON regardless of how the client sent it to us. With it, we'll re-encode config to native types. (@t0yv0 please sanity check me on this!)

**Comment by t0yv0** [comment]
Yeah that sounds right, I think for bridged providers what's persisted is not going to depend on how it's receiving the config, should be good there.

File: pkg/codegen/testing/test/testdata/transpiled_examples/cue-random-pp/dotnet/dotnet.csproj [hunk 3]
```diff
@@ -1,14 +1,15 @@
...
-    <TargetFramework>net7.0</TargetFramework>
-    <ImplicitUsings>enable</ImplicitUsings>
-    <Nullable>enable</Nullable>
-  </PropertyGroup>
-
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net6.0</TargetFramework>
```
Comment (iwahbe) [inline pkg/codegen/testing/test/testdata/transpiled_examples/cue-random-pp/dotnet/dotnet.csproj:5]: Why are we changing the target framework here?

**Review by iwahbe — APPROVED** [review]

File: pkg/codegen/testing/test/testdata/transpiled_examples/cue-random-pp/dotnet/dotnet.csproj (hunk 3 above)
Comment (blampe) [reply pkg/codegen/testing/test/testdata/transpiled_examples/cue-random-pp/dotnet/dotnet.csproj:5]: Probably a rebase side effect. This was bumped in #15373 and was rolled back when I ran `PULUMI_ACCEPT=true make test_codegen_dotnet`. I'm also not sure why it added `<PackageReference Include="Pulumi" Version="3.59.0" />`.

**Comment by t0yv0** [comment]
@EronWright brought up another interesting question on the strange Import(json=true) annotation in https://github.com/pulumi/pulumi-kubernetes/pull/3020

From what I can tell, Java SDK appears to generate this https://github.com/pulumi/pulumi-java/blob/main/pkg/codegen/java/gen.go#L449 to json-encode Provider configuration arguments specifically for explicit providers. This I think is an instance of this same problem.

**Comment by blampe** [comment]
> @EronWright brought up another interesting question on the strange Import(json=true) annotation in [pulumi/pulumi-kubernetes#3020](https://github.com/pulumi/pulumi-kubernetes/pull/3020)
> 
> From what I can tell, Java SDK appears to generate this https://github.com/pulumi/pulumi-java/blob/main/pkg/codegen/java/gen.go#L449 to json-encode Provider configuration arguments specifically for explicit providers. This I think is an instance of this same problem.

@t0yv0 yeah, https://github.com/pulumi/pulumi-java/pull/1308 covers the Java side. I'm almost inclined to suggest we start with Java (or a different language) if we don't have the appetite to change the behavior for everything at once.

**Comment by t0yv0** [comment]
Revisiting this PR, this could help remove a lot of difficulties in the bridge and even potentially remove config_encoding entirely, this is the right place to do the fix. @blampe @justinvp @mjeffryes what would it take to carry this forward? How can I help? 

Are there backwards compatibility concerns we can turn into test cases we can review? 

Starter list. To merge this PR:

- Do bridged providers correctly parse out naturally encoded CheckConfig/Confgiure calls after this PR in every case?
- Do native providers?
  - kubernetes
  - azure-native
  - go-provider based (I think covered above)

To remove config encoding parser bits entirely:
- After an upgrade, would engine ever call `prov_v2.Configure()` with the outputs of `prov_v1.CheckConfig()`?

**Comment by t0yv0** [comment]
Had a bad regression around this since bridge does not have tests for individual languages alas. Given this I'm adding this to M111 list see https://github.com/pulumi/pulumi-terraform-bridge/issues/2440

**Comment by t0yv0** [comment]
Continuing to iterate on this (https://github.com/pulumi/home/issues/3705), it's a bit more complicated.

**Comment by lbialy** [comment]
@t0yv0 any chances to get some insight on how this ended? We also have to serialize provider configs to json.

**Pull request closed without merging**